sdk/patina/src/arch/aarch64.rs

```diff
@@ -0,0 +1,129 @@
+//! AArch64 Specific abstractions for Patina.
+//!
+//! ## License
+//!
+//! Copyright (c) Microsoft Corporation.
+//!
+//! SPDX-License-Identifier: Apache-2.0
+//!
+//! Portions Copyright 2023 The arm-gic Authors.
+//! arm-gic is dual-licensed under Apache 2.0 and MIT terms.
+
+/// Reads and returns the value of the given aarch64 system register.
+#[macro_export]
+macro_rules! read_sysreg {
+    ($name:ident) => {
+        {
+            let mut value: u64;
+            // SAFETY: The caller must provide a valid system register name
+            // and ensure that the system is in a state where reading this register is safe.
+            unsafe {
+                core::arch::asm!(
+                    concat!("mrs {value:x}, ", core::stringify!($name)),
+                    value = out(reg) value,
+                    options(nomem, nostack),
+                );
+            }
+            value
+        }
+    }
+}
+
+/// Writes the given value to the given aarch64 system register.
+/// Usage:
+/// - `write_sysreg!(reg register_name, imm:value)` - Write immediate value
+/// - `write_sysreg!(reg register_name, imm:value, barrier)` - Write immediate value with barrier
+/// - `write_sysreg!(reg dest_register, reg src_register)` - Copy from one register to sysreg
+/// - `write_sysreg!(reg dest_register, reg src_register, barrier)` - Copy from one register to sysreg with barrier
+/// - `write_sysreg!(reg register_name, value)` - Write literal value
+/// - `write_sysreg!(reg register_name, value, barrier)` - Write literal value with barrier
+#[macro_export]
+macro_rules! write_sysreg {
```

Comment:
nit, I find it awkward that the barrier needs to be a string. You can fix that with this:

```rust
// OLD
(reg $dest:ident, imm $imm:literal, $barrier:expr) => {

// NEW
(reg $dest:ident, imm $imm:literal, $($barrier:tt)+) => {
```

And stringifying it looks like this: `core::stringify!($($barrier)+),`

Based off everything else it may even make sense to do this:

``` rust
(reg $dest:ident, imm $imm:literal, barrier $($barrier:tt)+) => {
```

Then your macro becomes 

```rust
//this
write_sysreg!(reg dbgwcr0_el1, value, isb sy);

// Or this if you use barrier
write_sysreg!(reg dbgwcr0_el1, value, barrier isb sy);
```

Reply:
Talked offline. Couldn't get the token tree to work with multiple barriers, kept string approach but allows multiple barriers.